src/common/telemetry/inc/Telemetry.h

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef TELEMETRY_H
+#define TELEMETRY_H
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include <dlfcn.h>
+#include <errno.h>
+#include <limits.h>
+#include <link.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include <version.h>
+
+#define TELEMETRY_BINARY_NAME "OSConfigTelemetry"
+#define TELEMETRY_TIMEOUT_SECONDS 10
+
+#define TELEMETRY_CORRELATIONID_ENVIRONMENT_VAR "activityId"
+#define TELEMETRY_RULECODENAME_ENVIRONMENT_VAR "_RuleCodename"
+
+// Helper macro to generate random filename
+#define generateRandomFilename() ({ char temp_template[] = "/tmp/telemetry_XXXXXX"; int fd = mkstemp(temp_template); char* filename_result = NULL; if (fd != -1) { close(fd); if (unlink(temp_template) == 0) { filename_result = (char*)malloc(strlen(temp_template) + 6); if (filename_result) { strcpy(filename_result, temp_template); strcat(filename_result, ".json"); } } } filename_result; })
+
+// Helper macro to get module directory
+#define getModuleDirectory() ({ Dl_info dl_info; char* directory_result = NULL; if (dladdr((void*)__func__, &dl_info) != 0) { if (dl_info.dli_fname != NULL) { const char* module_fullPath = dl_info.dli_fname; const char* lastSlash = strrchr(module_fullPath, '/'); if (lastSlash != NULL) { size_t dirLen = lastSlash - module_fullPath; directory_result = (char*)malloc(dirLen + 1); if (directory_result) { strncpy(directory_result, module_fullPath, dirLen); directory_result[dirLen] = '\0'; } } } } directory_result; })
+
+// Buffer sizes for string conversion of numeric values
+// Based on maximum possible digits for each type plus sign and null terminator
+#define MAX_INT_STRING_LENGTH 16    // Accommodates 32-bit int values
+#define MAX_LONG_STRING_LENGTH 32   // Accommodates 64-bit long values
+
+// Static file handle for telemetry logging with thread safety
+static FILE* g_telemetryFile __attribute__((unused)) = NULL;
+static char* g_fileName __attribute__((unused)) = NULL;
+static char* g_moduleDirectory __attribute__((unused)) = NULL;
+static int g_telemetryFileInitialized __attribute__((unused)) = 0;
+
+#ifdef DEBUG
+#define VERBOSE_FLAG_IF_DEBUG "-v"
+#else
+#define VERBOSE_FLAG_IF_DEBUG ""
+#endif
+
+// Initialize telemetry file - call once at program start (thread-safe)
+#define OSConfigTelemetryInit() do { if (!g_telemetryFileInitialized) { static int initLock = 0; if (__sync_bool_compare_and_swap(&initLock, 0, 1)) { if (!g_telemetryFile) { g_moduleDirectory = getModuleDirectory(); g_fileName = generateRandomFilename(); g_telemetryFile = fopen(g_fileName, "a"); if (g_telemetryFile) { g_telemetryFileInitialized = 1; } } } else { while (!g_telemetryFileInitialized) { usleep(1000); } } } } while(0)
```

Comment:
>usleep(1000); [](http://example.com/codeflow?start=403&length=13)

No waits, please. As stated, when the entire audit can take less than 1 second, we should block here with seconds, possibly several

This is blocking